File: finanze/domain/calculations.py
from __future__ import annotations

from enum import Enum
from typing import List, Optional

from domain.dezimal import Dezimal
from pydantic.dataclasses import dataclass


class SavingsPeriodicity(str, Enum):
    MONTHLY = "MONTHLY"
    QUARTERLY = "QUARTERLY"
    YEARLY = "YEARLY"

    @property
    def periods_per_year(self) -> int:
        if self is SavingsPeriodicity.MONTHLY:
            return 12
        if self is SavingsPeriodicity.QUARTERLY:
            return 4
        return 1


@dataclass
class SavingsScenarioRequest:
    scenario_id: str
    annual_market_performance: Dezimal
    periodic_contribution: Optional[Dezimal] = None
    target_amount: Optional[Dezimal] = None


@dataclass
class SavingsRetirementRequest:
    withdrawal_amount: Optional[Dezimal] = None
    withdrawal_years: Optional[int] = None


@dataclass
class SavingsCalculationRequest:
    base_amount: Optional[Dezimal]
    years: Optional[int]
    periodicity: SavingsPeriodicity
    scenarios: List[SavingsScenarioRequest]
    retirement: Optional[SavingsRetirementRequest] = None


@dataclass
class SavingsPeriodEntry:
    period_index: int
    contributed: Dezimal
    total_contributed: Dezimal
    revaluation: Dezimal
    total_revaluation: Dezimal
    total_invested: Dezimal
    balance: Dezimal


@dataclass
class SavingsRetirementPeriodEntry:
    period_index: int
    withdrawal: Dezimal
    total_withdrawn: Dezimal
    revaluation: Dezimal
    balance: Dezimal


@dataclass
class SavingsScenarioResult:
    scenario_id: str
    annual_market_performance: Dezimal
    periodic_contribution: Dezimal
    accumulation_periods: List[SavingsPeriodEntry]
    total_contributions: Dezimal
    total_revaluation: Dezimal
    final_balance: Dezimal
    retirement: Optional[SavingsRetirementResult]


@dataclass
class SavingsRetirementResult:
    withdrawal_amount: Dezimal
    duration_periods: int
    duration_years: Dezimal
    total_withdrawn: Dezimal
    periods: List[SavingsRetirementPeriodEntry]


@dataclass
class SavingsCalculationResult:
    scenarios: List[SavingsScenarioResult]

File: mobile/src/domain/core/calculations.ts
import { Dezimal } from "@/domain/dezimal"

export enum SavingsPeriodicity {
  MONTHLY = "MONTHLY",
  QUARTERLY = "QUARTERLY",
  YEARLY = "YEARLY",
}

export interface SavingsScenarioRequest {
  scenarioId: string
  annualMarketPerformance: Dezimal
  periodicContribution?: Dezimal | null
  targetAmount?: Dezimal | null
}

export interface SavingsRetirementRequest {
  withdrawalAmount?: Dezimal | null
  withdrawalYears?: number | null
}

export interface SavingsCalculationRequest {
  baseAmount: Dezimal | null
  years: number | null
  periodicity: SavingsPeriodicity
  scenarios: SavingsScenarioRequest[]
  retirement?: SavingsRetirementRequest | null
}

export interface SavingsPeriodEntry {
  periodIndex: number
  contributed: Dezimal
  totalContributed: Dezimal
  revaluation: Dezimal
  totalRevaluation: Dezimal
  totalInvested: Dezimal
  balance: Dezimal
}

export interface SavingsRetirementPeriodEntry {
  periodIndex: number
  withdrawal: Dezimal
  totalWithdrawn: Dezimal
  revaluation: Dezimal
  balance: Dezimal
}

export interface SavingsScenarioResult {
  scenarioId: string
  annualMarketPerformance: Dezimal
  periodicContribution: Dezimal
  accumulationPeriods: SavingsPeriodEntry[]
  totalContributions: Dezimal
  totalRevaluation: Dezimal
  finalBalance: Dezimal
  retirement: SavingsRetirementResult | null
}

export interface SavingsRetirementResult {
  withdrawalAmount: Dezimal
  durationPeriods: number
  durationYears: Dezimal
  totalWithdrawn: Dezimal
  periods: SavingsRetirementPeriodEntry[]
}

export interface SavingsCalculationResult {
  scenarios: SavingsScenarioResult[]
}
